packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts

```diff
@@ -0,0 +1,257 @@
+import { Server } from "http";
+import { Server as SecureServer } from "https";
+
+import type { Express } from "express";
+import { Optional } from "typescript-optional";
+
+import {
+  Logger,
+  Checks,
+  LogLevelDesc,
+  LoggerProvider,
+} from "@hyperledger/cactus-common";
+import {
+  ICactusPlugin,
+  ICactusPluginOptions,
+  IPluginWebService,
+  IWebServiceEndpoint,
+  PluginAspect,
+} from "@hyperledger/cactus-core-api";
+
+import { KeyVaultSecret, SecretClient } from "@azure/keyvault-secrets";
+import {
+  UsernamePasswordCredential,
+  DefaultAzureCredential,
+} from "@azure/identity";
+
+import { PluginKeychainAzureKvMock } from "../../test/typescript/mock/plugin-keychain-azure-kv-mock";
+
+// TODO: Writing the getExpressRequestHandler() method for
+// GetKeychainEntryEndpointV1 and SetKeychainEntryEndpointV1
+// import { GetKeychainEntryEndpointV1 } from "./web-services/get-keychain-entry-endpoint-v1";
+// import { SetKeychainEntryEndpointV1 } from "./web-services/set-keychain-entry-endpoint-v1";
+
+export enum AzureCredentialType {
+  LocalFile = "LOCAL_FILE",
+  InMemory = "IN_MEMORY",
+}
+
+export interface IAzureInMemoryCredentials {
+  azureTenantId: string;
+  azureClientId: string;
+  azureUsername: string;
+  azurePassword: string;
+}
+
+export interface IPluginKeychainAzureKvOptions extends ICactusPluginOptions {
+  logLevel?: LogLevelDesc;
+  keychainId: string;
+  instanceId: string;
+  azureEndpoint: string;
+  azureCredentialType: AzureCredentialType;
+  azureInMemoryCredentials?: IAzureInMemoryCredentials;
+}
+
+export class PluginKeychainAzureKv implements ICactusPlugin, IPluginWebService {
+  public static readonly CLASS_NAME = "PluginKeychainAzureKv";
+
+  readonly vaultUrl: string;
+  private readonly log: Logger;
+  private readonly instanceId: string;
+  private endpoints: IWebServiceEndpoint[] | undefined;
+  private azureKvClientMock: PluginKeychainAzureKvMock;
```

Comment:
@jagpreetsinghsasan Code in the ./src/main/... directory cannot import anything from the ./src/test/... directory. The trick is to rename this to `azureKvClient` (or `backend` if you want) and declare it's type as `SecretClient`.
Then add the same thing as a parameter to the constructor and for the tests pass in an instance of your mock class.

TLDR: There can be no trace of any mock types nor mock implementations in this class, it has to be abstracted out.
The way it's written now AFAICT if I deployed this to production today it would just be running the mock instead of using the real `SecretClient`.

Reply:
I have renamed this variable to azureKvClient, but in the plugin class constructor, we have this
![image](https://user-images.githubusercontent.com/56861721/119935369-050a4400-bfa5-11eb-9210-05baa424a76e.png)

What should I be doing about this? (Because for this initialization, we need to import the mock class as well)

Reply:
@jagpreetsinghsasan I'd extend the constructor parameters to have not just the azureCredentials, but also a `backend?: SecretClient` parameter. Then in the constructor I'd check if `options.backend` is truthy and if yes, I'd use that and ignore the azureCredentials.
This would allow the tests to instantiate the plugin with the mock and without the plugin needing to know anything about the mock or contain any implementation that's for the mock.